src/systemctl/systemctl.c

```diff
@@ -5605,6 +5605,51 @@ static int mangle_names(char **original_names, char ***mangled_names) {
         return 0;
 }
 
+static int unit_exists(const char *unit) {
+        _cleanup_(sd_bus_message_unrefp) sd_bus_message *reply = NULL, *m = NULL;
+        _cleanup_(sd_bus_error_free) sd_bus_error error = SD_BUS_ERROR_NULL;
+        _cleanup_free_ char *path = NULL;
+        static const struct bus_properties_map property_map[] = {
+                { "LoadState",   "s", NULL, offsetof(UnitStatusInfo, load_state)  },
+                { "ActiveState", "s", NULL, offsetof(UnitStatusInfo, active_state)},
+                {},
+        };
+        UnitStatusInfo info = {};
+        sd_bus *bus;
+        int r;
+
+        path = unit_dbus_path_from_name(unit);
+        if (!path)
+                return log_oom();
+
+        r = acquire_bus(BUS_MANAGER, &bus);
+        if (r < 0)
+                return r;
+
+        r = sd_bus_call_method(
+                        bus,
+                        "org.freedesktop.systemd1",
+                        path,
+                        "org.freedesktop.DBus.Properties",
+                        "GetAll",
+                        &error,
+                        &reply,
+                        "s", "");
+        if (r < 0)
+                return log_error_errno(r, "Failed to get properties: %s", bus_error_message(&error, r));
+
+        r = bus_message_map_all_properties(reply, property_map, &info);
+        if (r < 0)
+                return log_error_errno(r, "Failed to map properties: %s", bus_error_message(&error, r));
+
+        if (streq_ptr(info.load_state, "not-found") && streq_ptr(info.active_state, "inactive")) {
+                log_notice("Unit %s does not exist.", unit);
+                return 0;
+        }
+
+        return 1;
+}
```

Comment:
please make unit_exists() return the result of the streq_ptr() checks, and let the caller do the log_notice() invocation.

i.e. replace the if block with just this:

{{{c
…
return !streq_ptr(info.load_state, "not-found") || !streq_ptr(info.active_state, "inactive");
…
}}}

That way, the function returns negative on error, "0" if the unit doesn't exist, and "1" if it does exist.